common.c

```diff
@@ -35,40 +36,104 @@ struct addrinfo *get_address(const char *host, const char *port)
     }
 }
 
+#ifdef __linux__
+    #ifndef UDP_SEGMENT
+        #define UDP_SEGMENT 103
+    #endif
+#endif
+
+bool send_dgrams_default(int fd, struct sockaddr *dest, struct iovec *dgrams, size_t num_dgrams)
+{
+    for(size_t i = 0; i < num_dgrams; ++i) {
+        struct msghdr mess = {
+            .msg_name = dest,
+            .msg_namelen = quicly_get_socklen(dest),
+            .msg_iov = &dgrams[i], .msg_iovlen = 1
+        };
+
+        ssize_t bytes_sent;
+        while ((bytes_sent = sendmsg(fd, &mess, 0)) == -1 && errno == EINTR);
+        if (bytes_sent == -1) {
+            perror("sendmsg failed");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool send_dgrams_gso(int fd, struct sockaddr *dest, struct iovec *dgrams, size_t num_dgrams)
+{
+    struct iovec vec = {
+        .iov_base = (void *)dgrams[0].iov_base,
+        .iov_len = dgrams[num_dgrams - 1].iov_base + dgrams[num_dgrams - 1].iov_len - dgrams[0].iov_base
+    };
+
+    struct msghdr mess = {
+        .msg_name = dest,
+        .msg_namelen = quicly_get_socklen(dest),
+        .msg_iov = &vec,
+        .msg_iovlen = 1
+    };
+
+    union {
+        struct cmsghdr hdr;
+        char buf[CMSG_SPACE(sizeof(uint16_t))];
+    } cmsg;
+    if (num_dgrams != 1) {
+        cmsg.hdr.cmsg_level = SOL_UDP;
+        cmsg.hdr.cmsg_type = UDP_SEGMENT;
+        cmsg.hdr.cmsg_len = CMSG_LEN(sizeof(uint16_t));
+        *(uint16_t *)CMSG_DATA(&cmsg.hdr) = dgrams[0].iov_len;
+        mess.msg_control = &cmsg;
+        mess.msg_controllen = (socklen_t)CMSG_SPACE(sizeof(uint16_t));
+    }
+
+    ssize_t bytes_sent;
+    while ((bytes_sent = sendmsg(fd, &mess, 0)) == -1 && errno == EINTR);
+    if (bytes_sent == -1) {
+        perror("sendmsg failed");
+        return false;
+    }
+
+    return true;
+}
+
+bool (*send_dgrams)(int fd, struct sockaddr *dest, struct iovec *dgrams, size_t num_dgrams) = send_dgrams_default;
```

Comment:
The code is easier to read if the `send_dgrams` definition is split up into a typedef for the pointer type and a separate variable declaration. Moreover we should move this to the top of the header, as it is a global variable and add a comment that this is a setting, hence the global variable. I believe it's often helpful to document the reason for a global variable. We should also make the variable static.

Reply:
This follows the structure as implemented in quicly cli: https://github.com/h2o/quicly/blob/master/src/cli.c#L403-L461
While I agree with you, it might be better to follow the quicly implementation for consistency

Reply:
Ok, that makes sense